src/frontend/src/optimizer/plan_node/generic/join.rs

```diff
@@ -213,76 +213,68 @@ impl<PlanRef: GenericPlanRef> GenericPlanNode for Join<PlanRef> {
         let full_out_col_num = self.internal_column_num();
         let i2o = ColIndexMapping::with_remaining_columns(&self.output_indices, full_out_col_num);
 
-        let mut pk_indices = left_pk
+        // Collect PKs in internal column space (without applying i2o mapping yet)
+        let mut pk_indices_internal = left_pk
             .iter()
             .map(|index| l2i.try_map(*index))
             .chain(right_pk.iter().map(|index| r2i.try_map(*index)))
             .flatten()
-            .map(|index| i2o.try_map(index))
-            .collect::<Option<Vec<_>>>()?;
```

Comment:
The derivation of join keys logic is changed. We apply the `i2o` mapping at the end. Otherwise, index selection for streaming join might panic, because index selection will replace a primary table with an index. Before this PR, they will have the same stream keys, but after this PR, they have different stream keys. For other operator like Agg, TopN, we can safely replace it with indexes, because the index stream_key's existence can always be  guaranteed after `logical_rewrite_for_stream`. However, for the join, we only keep one side of the join key as the join's stream_key, and the other side's join key could be removed (even if it is the stream_key of the index).